docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+The list of affected zones could be maintained in the [`Message`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L168).
+
+##### Possible Actions
+
+- If necessary, `kube-controller-manager` could be scaled down as described [above](#possible-actions).
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- The `machine-controller-manager` could be either scaled down as described [above](#possible-actions) or be enhanced to avoid (or ramp down) deletion of machines from the affected zones (or `machinedeployments`).
+- Gardener can defer any disruptive updates until later.
+
+### Condition Types of future relevance
+
+This is just a sample set of condition types that might be of future relevance. It is not an exhaustive list.
+The suggested possible actions for these condition types are also just samples and not an exhaustive list.
+
+#### InternalNetworkHealthy
+
+If there is a way to detect problems in the cluster network that is severe enough to affect any significant portion of the cluster then this could be encoded in a component that updates the conditions with the `InternalNetworkHealthy` condition type.
+
+##### Possible Actions
+
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- Gardener can defer any disruptive updates until later.
+
+#### CloudProviderHealthy
+
+If there is information about some issues reported against the cloud provider, it could be updated in the conditions with the `CloudProviderHealthy` condition type.
+If there is more specific information about the issue then more specific condition types can be defined.
+
+Such conditions could be updated manually by the administrators, operators on duty based on publications of such issues by the cloud providers.
+If there is a way to automatically scrape such information from well-known locations or APIs then this could even be automated.
+Such conditions need not be manually maintained in every single `Cluster` resource instance for each shoot cluster. We can leverage the advantages of operator design-pattern and build on this basic functionality by introducing higher order resources (such as to capture the cloud provider health conditions) and propagate the such conditions to all the relevant `Cluster` resources automatically.
+
+##### Possible Actions
+
+- Controllers such as `kube-controller-manager`, `machine-controller-manager` that manage crucial resources such as `nodes` can be either scaled down as describe [above](#possible-actions) or enhanced to change their behaviour to ramp down their active interventions to the cluster.
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- Gardener can defer any disruptive updates until later.
+
+#### CloudProviderQuotaExceeded
+
+If there is a way to detect that some cloud provider quotas are exceeded this information can be updated in the conditions with the `CloudProviderQuotaExceeded` condition type.
```

Comment:
What we will achieve throttling/rate limiting the end user resources on `CloudProviderQuotaExceeded`? 
How you plan to achieve this throttling/rate-limiting?

Reply:
> What we will achieve throttling/rate limiting the end user resources on CloudProviderQuotaExceeded?

Fail fast.

> How you plan to achieve this throttling/rate-limiting?

One way is by introducing a reserve proxy in front of the `kube-apiserver`? There might be other ways.

Reply:
Isn't `CloudProviderQuotaExceeded` too generic? Is this Condition also designed for Shoots or only for seeded Shoots? The quota might be from any type. How you decide which resources to throttle?
What we win (expect fail fast) throttling on `CloudProviderQuotaExceeded`? Does it worth the complexity for only fail fast?

Reply:
> Isn't CloudProviderQuotaExceeded too generic?
Yes. Unless we use `Reason` to add more specific detail. A more specific condition type also is an option.

> Is this Condition also designed for Shoots or only for seeded Shoots?

Both.

> How you decide which resources to throttle?

Would depend on which quota is exceeded. For example, if loadbalancer quota is exceeded, `LodBalancer` `services` could be throttled. Could even be done via a validating webhook instead of a reverse proxy.

> What we win (expect fail fast) throttling on CloudProviderQuotaExceeded? Does it worth the complexity for only fail fast?

Fail fast is always desirable unless the requesting component is a controller itself which monitors status of the resource it is provisioning actively and intelligently.